src/common/proxy_wasm/ngx_proxy_wasm_properties.c

```diff
@@ -877,25 +879,58 @@ ngx_proxy_wasm_properties_get_host(ngx_proxy_wasm_ctx_t *pwctx,
 
     hpn = (host_props_node_t *)
               ngx_str_rbtree_lookup(&pwctx->host_props_tree, path, hash);
+    if (!hpn) {
+        return NGX_DECLINED;
+    }
 
-    if (hpn) {
-        value->data = hpn->value.data;
-        value->len = hpn->value.len;
+    /* stored value is not const: need to call getter again */
+    if (!hpn->is_const && pwctx->host_prop_getter) {
+        return NGX_DECLINED;
+    }
 
-        return NGX_OK;
+    value->data = hpn->value.data;
+    value->len = hpn->value.len;
+
+    if (hpn->negative_cache) {
+        return NGX_BUSY;
     }
 
-    return NGX_DECLINED;
+    return NGX_OK;
 }
 
 
-static ngx_int_t
+static ngx_inline ngx_int_t
+set_hpn_value(host_props_node_t *hpn, ngx_str_t *value, unsigned is_const,
+    ngx_pool_t *pool)
+{
+    unsigned char  *new_data;
+
+    if (value->data == NULL) {
+        new_data = NULL;
+        hpn->negative_cache = 1;
```

Comment:
Isn't `negative_cache` supposed to only be for consts?

Reply:
yes, but for non-consts this function is never called.